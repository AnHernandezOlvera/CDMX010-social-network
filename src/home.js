// target es un parametro arbitrario que siempre recibe root/rootDIv
import { getAllPosts } from './lib/firebase.js';

export const home = async (target) => {
  const templeteHome = `
    <div id="back-list"><div id="post-list"></div>
`;
  // eslint-disable-next-line no-param-reassign
  target.innerHTML = templeteHome;
  const posts = await getAllPosts();
  const postTemplates = posts.map((post) => `
    <a href="#" class="single-post" data-id="${post.id}">
      <h2 class="title-list">${post.title}</h2>
    </a>
  `);

  const postList = document.getElementById('post-list');
  postList.innerHTML = postTemplates.join('');

  const btnsPost = document.querySelectorAll('.single-post');
  btnsPost.forEach((btn) => {
    btn.addEventListener('click', async (e) => {
      e.preventDefault();
      // onNavigate('/singlepost');
      // const doc = await getPostInfo(e.target.dataset.id);
      // console.log(doc);
    });
  });
};

export default home;
