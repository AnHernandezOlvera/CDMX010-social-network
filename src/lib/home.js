const home = `
    <div id="back-list"><div id="post-list"></div>
`;
