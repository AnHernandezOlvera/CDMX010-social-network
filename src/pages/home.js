export const home = `
<h1>Soy home</h1>
`;
export default home;
