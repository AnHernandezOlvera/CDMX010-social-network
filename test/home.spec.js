// importamos funciones que nos ayudaran a simular las interacciones del usario con la UI
import { home } from '../src/home.js';


// importamos funciones que nos ayudaran a simular las interacciones del usario con la UI
import { screen, waitFor } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';


describe('home', () => {
  // Creamos un root element en nuestro DOM para que nos sirva de contenedor para nuesta vista/componente principal Login
  beforeEach(() => {
    document.body.innerHTML = '<div id="root"></div>';
  });

  // Este test puede ser saltado enfocate en los 2 de aabjo
  it('should render', async () => {
    const target = document.getElementById('root');
    const getAllPosts = jest.fn().mockImplementation(() => Promise.resolve([
      { id: '173763276327323', title: 'test'}
    ]));
    const firebase = { getAllPosts };
    await home(target, firebase);

    expect(target.innerHTML).toMatchSnapshot();
  });

  it.skip('should login the user when the user submits the login form', () => {
    const target = document.getElementById('root');
    
    // Mockeamos la funcion de login de Firebase la cual sera invocada cuando el usuario haga click sobre el boton "Sign In"
    // Por que?  La idea central es verificar que esta funcion es invocada lo cual haremos linea mas abajo
    const signIn = jest.fn().mockImplementation(() => Promise.resolve('ok'));
    
    const email = 'test@example.com';
    const password = '123456789';
    
    // Invoco mi component y lo renderizo pasandole el contenedor como argumento
    Login({ signIn }).render(target);

    // Referencia: https://testing-library.com/docs/ecosystem-user-event/#api
    userEvent.type(screen.getByLabelText('Email'), email);
    userEvent.type(screen.getByLabelText('Password'), password);
    userEvent.click(screen.getByDisplayValue('Sign In'));
    
    // Verificamos que nuestro mock de la funcion login de firebase es llamada con el email y password que el usuario escribio mas arriba
    // Referencia: https://jestjs.io/docs/en/expect#tohavebeencalledwitharg1-arg2-
    expect(signIn).toHaveBeenCalledWith(email, password);
  });

  it.skip('should show an error message when a user provides invalid credentials', async () => {
    const target = document.getElementById('root');
    const message = 'Invalid credentials!';
    const signIn = jest.fn().mockImplementation(
      (email, password) => Promise.reject({ message }),
    );
    const email = 'test@example.com';
    const password = 'invalidpassword';
    Login({ signIn }).render(target);

    // Referencia: https://testing-library.com/docs/ecosystem-user-event/#api
    userEvent.type(screen.getByLabelText('Email'), email);
    userEvent.type(screen.getByLabelText('Password'), password);
    userEvent.click(screen.getByDisplayValue('Sign In'));
    
    // Referencia: https://jestjs.io/docs/en/expect#tohavebeencalledwitharg1-arg2-
    expect(signIn).toHaveBeenCalledWith(email, password);

    // Referencia: https://testing-library.com/docs/dom-testing-library/api-async/#waitfor
    await waitFor(() => {
      expect(screen.getByTestId('errorMessage').innerHTML).toBe(message);
    });
  });
});