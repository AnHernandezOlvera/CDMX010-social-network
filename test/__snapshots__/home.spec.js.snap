// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home should render 1`] = `
"
    <div id=\\"post-list\\">
  <div class=\\"card\\">
    <a href=\\"/singlepost?id=173763276327323\\" class=\\"single-post\\" data-id=\\"173763276327323\\">
      <h2 class=\\"title-list\\" data-id=\\"173763276327323\\">test</h2>
    </a> 
  </div>
  </div>
    
<footer>
    <div id=\\"iconsFooter\\" class=\\"iconsfooter\\">
    <span><a id=\\"toHome\\" href=\\"#\\"><i class=\\"fas fa-arrow-alt-circle-left\\"></i></a></span>
    <span><a id=\\"toPosting\\" href=\\"#\\"> <i class=\\"fas fa-plus-circle\\"></i></a></span>
    <span><a id=\\"signOut\\" href=\\"#\\"><i class=\\"fas fa-sign-out-alt\\"></i></a></span>
    </div>
  </footer>

"
`;
